### PR TITLE
HOTFIX: remove from restoringByPartition once restored

### PR DESCRIPTION
**Pull request by ableegoldman**
Minor follow up to https://github.com/apache/kafka/pull/7608:
For some reason the `AssignedStreamTasks#updateRestored` method only updates the `restoring` and `restoredPartitions` data structures, but there is a third map holding restored tasks & partitions: `restoringByPartitions` -- this was causing failures during `onPartitionsLost` due to the new strict emptiness checks introduced in https://github.com/apache/kafka/pull/7608)

Also fixes a strict emptiness check based on a faulty assumption (introduced in https://github.com/apache/kafka/pull/7608) , as the `partitionInfo` in `StoreChangelogReader` is just the latest metadata for all topics and does not need to be, and shouldn't be, empty.

Also improves the `TaskManager#closeLostTasks` logging, by separating by case and logging the specific failure before throwing.

Needs to be cherry-picked to 2.4 as well

### CODE REVIEW

**Comment by ableegoldman** [comment]
System tests don't seem to be building any time soon but once they do you can find them here: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3358

Note that on only 1 repeat we are seeing multiple failures before this, so that should be sufficient to verify.

@guozhangwang @abbccdda

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 1]
```diff
@@ -361,6 +363,8 @@ void updateRestored(final Collection<TopicPartition> restored) {
         }
         if (allTasksRunning()) {
             restoredPartitions.clear();
+            restoringByPartition.clear();
+            restoring.clear();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Why is this necessary? In the `for` loop above, each partition that finished restoring should be removed vai `it.remove();` -- hence, if `allTasksRunning()` return `true` `restoring` should be empty ?

Similar for `restoringByPartition` given the two new lines this PR adds.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 2]
```diff
@@ -505,11 +509,17 @@ public void shutdown(final boolean clean) {
...
     public boolean isEmpty() {
-        return super.isEmpty()
-            && restoring.isEmpty()
-            && restoringByPartition.isEmpty()
-            && restoredPartitions.isEmpty()
-            && suspended.isEmpty();
+        if (restoring.isEmpty() && !restoringByPartition.isEmpty()) {
+            log.error("Assigned stream tasks in an inconsistent state: the set of restoring tasks is empty but the " +
+                      "restoring by partitions map contained {}", restoringByPartition);
+            return false;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: It is save to return `false`? Should we rather throw an `IllegalStateException` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java [hunk 3]
```diff
@@ -183,9 +183,15 @@ void clear() {
     }
 
     boolean isEmpty() {
-        return runningByPartition.isEmpty()
-                   && running.isEmpty()
-                   && created.isEmpty();
+        if (running.isEmpty() && !runningByPartition.isEmpty()) {
+            log.error("Assigned stream tasks in an inconsistent state: the set of running tasks is empty but the " +
+                          "running by partitions map contained {}", runningByPartition);
+            return false;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java]: As above

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 4]
```diff
@@ -269,10 +269,16 @@ void closeRevokedSuspendedTasks() {
 
         if (exception != null) {
             throw exception;
-        } else if (!(active.isEmpty() && assignedActiveTasks.isEmpty() && changelogReader.isEmpty())) {
+        } else if (!(active.isEmpty())) {
+            log.error("The set of active tasks was non-empty: {}", active);
+            throw new IllegalStateException("TaskManager found leftover active task state after closing all zombies");
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: If we throw there, we don't check the other conditions and might miss some logs. Might be better to only log for each condition and throw only after each condition was checked (and if at least one condition hit)?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Btw: we maintain state quite redundantly (what is a general issue). Would it make sense to extract a helper class that wraps
```
private final Map<TaskId, StreamTask> restoring = new HashMap<>();
private final Set<TopicPartition> restoredPartitions = new HashSet<>();
private final Map<TopicPartition, StreamTask> restoringByPartition = new HashMap<>();
```

and we only modify the state via the helper? The helper class can insure that all three data structures don't diverge. (Also ok to do in a follow up cleanup PR)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: +1, we should just throw in this case.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Ack to the first point, we shouldn't need to clear these here. I also completely agree with the second comment, but we want to limit these PRs to just the necessary fixes for now -- we've been discussing some other large-scale refactoring and cleanup of this and related classes on other PRs and tried to drop most of the thoughts on this ticket, if you have anything else to add: https://issues.apache.org/jira/browse/KAFKA-9113

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Well, yeah we should

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 4 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Yeah actually that would be significantly more helpful for debugging, I'll make it happen

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Oh right I remember why I didn't want to just throw outright here, because we want to check all three objects for emptiness in `TaskManager#closeLostTasks` and then log which weren't empty, and what was in them (as you commented below). And it would be awkward if the `active.isEmpty` condition itself threw, since we're testing that condition to determine if we need to throw a IllegalStateException in the first place.

Does that make sense?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 5]
```diff
@@ -160,15 +163,16 @@ private void addNewActiveTasks(final Map<TaskId, Set<TopicPartition>> newActiveT
     }
 
     private void addNewStandbyTasks(final Map<TaskId, Set<TopicPartition>> newStandbyTasks) {
-        log.trace("New standby tasks to be created: {}", newStandbyTasks);
+        log.debug("New standby tasks to be created: {}", newStandbyTasks);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:166]: Not sure why this was trace as the same message is debug in the parallel active method

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: I do see the motivation for throwing, since the `isEmpty` methods may be re-used in the future for cases other than throwing an `IllegalStateException` anyways. 
On the other hand, it really is awkward to throw an exception from an `isEmpty` conditional and I'm not sure a future user of this method would be expecting it to throw an exception, or notice that it does... I'm still open to throwing, just laying out my main concern

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: I did change it to throw though, so if you're satisfied then we can merge -- once the final system test run passes so we don't go through this again :) 

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3363/

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 5 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:166]: Good catch!

**Review by mjsax — APPROVED** [review]
LGTM, assuming Jenkins passes.

**Comment by ableegoldman** [comment]
@mjsax The last system test had a failure but I think that was kicked off between patches, ie I changed something then kicked it off then fixed it up further -- but I kicked off more tests to confirm. I'll ping when the system tests are all green

**Comment by ableegoldman** [comment]
@mjsax @guozhangwang 80/80 total system test runs were green (kicked off quite a few to be safe):

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3365
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3366
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3367
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3368
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3369

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk and cherry-picked to 2.4.
